### PR TITLE
refactor isomorphic tests

### PR DESCRIPTION
**Pull request by giladgray**
cc @codebling 

#### Changes proposed in this pull request:

- DRY isotest code by introducing helper function that does all the work
- move usage into `test/` directory in each package

### CODE REVIEW

**Comment by giladgray** [comment]
@codebling how can i be sure that these tests are actually _passing_ and not reporting false positives? can you describe an easy way to make a component fail the test? (i'm an SSR noob and could use a hand.)

File: test/isotest.js [hunk 1]
```diff
@@ -21,23 +21,19 @@ const isReactClass = (Component) => {
...
+ * @param customProps       {{ [componentName: string]: any}}               custom props per component
+ * @param customChildren    {{ [componentName: string]: React.ReactNode }}  custom children per component
+ */
+module.exports = function isotest(Components, customProps, customChildren) {
+    Object.keys(Components).forEach((ComponentKey) => {
         const Component = Components[ComponentKey];
-
         if (isReactClass(Component)) {
-            it("can be server-rendered", () => {
+            it(`<${ComponentKey}> can be server-rendered`, () => {
```
Comment (giladgray) [inline test/isotest.js]: moved component name into test name so they can be easily distinguished when run in parallel (previously all `it`s had the exact same name)

**Comment by giladgray** [comment]
also @codebling is it _isometric_ or _isomorphic_? I think the latter is correct, as in [`isomorphic-fetch`](https://github.com/matthew-andrews/isomorphic-fetch)

**Comment by adidahiya** [comment]
✅  isomorphic

**Comment by codebling** [comment]
@giladgray you can fail a test by attempting simply including a reference to `document` in the constructor (or any of the life cycle methods that run before `DidMount`, which never gets called on the server). `document` is undefined server-side

File: gulp/aliases.js [hunk 2]
```diff
@@ -22,7 +22,8 @@ module.exports = (gulp) => {
     gulp.task("build", (done) => rs("clean", "compile", "webpack-compile-docs", done));
 
     // build code, run unit tests, terminate
-    gulp.task("test", ["test-dist", "karma", "isotest"]);
+    // NOTE: circle.yml runs each of these separately for better error reporting; be sure to sync
```
Comment (adidahiya) [inline gulp/aliases.js]: "be sure to sync"? what does that mean? also, how does it produce better error reporting? does it run the next gulp task even if a previous one fails?

File: test/isotest.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    ;
+};
+
+/**
+ * Tests that each ComponentClass in Components can be isomorphically rendered on the server.
+ * @param Components        {{ [componentName: string]: any }}              main export from package
+ * @param customProps       {{ [componentName: string]: any}}               custom props per component
+ * @param customChildren    {{ [componentName: string]: React.ReactNode }}  custom children per component
+ */
+module.exports = function isotest(Components, customProps, customChildren) {
```
Comment (adidahiya) [inline test/isotest.js]: can this be more verbose to make it clear that it _generates_ multiple tests? `generateIsoTests`? same for the import name elsewhere

File: gulp/util/karma-config.js [hunk 4]
```diff
@@ -54,12 +54,15 @@ module.exports = function createConfig (project) {
...
             watermarks: {
-                lines: [80, 90],
-                statements: [80, 90],
+                lines: [79, 90],
+                statements: [79, 90],
             },
         },
         files: filesToInclude,
         frameworks: ["mocha", "chai", "phantomjs-shim", "sinon"],
+        mime: {
```
Comment (adidahiya) [inline gulp/util/karma-config.js]: curious, why do we need this now?

**Comment by giladgray** [comment]
@codebling got it. i tried messing with a few components and the tests didn't fail so i tried using enzyme's static HTML renderer and it revealed a handful of problematic components. 

see the [latest build results](https://circleci.com/gh/palantir/blueprint/1503?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) for some legit errors.

File: gulp/util/karma-config.js (hunk 4 above)
Comment (giladgray) [inline gulp/util/karma-config.js]: it was added in #409, then lost when in #381 (notice the smaller number 😄), and now restored here.

File: gulp/aliases.js (hunk 2 above)
Comment (giladgray) [inline gulp/aliases.js]: means keep the two uses in sync. make sure they run the same tasks.

https://circleci.com/gh/palantir/blueprint/1503 shows two of the three failing.

i have an idea how to post comments on build failure, and we could post a different comment based on which ones fail.

**Comment by adidahiya** [comment]
please file follow-up tickets for any skipped tests

**Comment by codebling** [comment]
@giladgray I see only passing tests in the build you linked. As far as switching to the Enzyme renderer, if you're able to find a way to pass specific Props and Children, then it should work. I wasn't able to find a way to do that with Enzyme (if I recall correctly), which is why I switched to ReactTestUtils. It's explained in the first comment in #381.

File: test/isotest.js (hunk 1 above)
Comment (codebling) [inline test/isotest.js]: I haven't had a look at all of the changes yet, but in my experience, console outputs can become interlaced when running in parallel/asynchronously. I'd imagined this to be the reason why the tests were set up to run in series when I first found then.

**Comment by giladgray** [comment]
@codebling ah you're right that link was wrong. i fixed it above. here's the correct build:
https://circleci.com/gh/palantir/blueprint/1508.

regarding the specific children: you'd already built all the pieces with `customProps` and `customChildren`!  :trophy: I just [filled them out](https://github.com/palantir/blueprint/pull/595/files#diff-8ee2818aafe494f21fc1a0effbf51aebR19). but that validations happens in the constructor, which I believe is supposed to run as part of the server render? it seemed to me like `TestUtils` wasn't doing enough work, or wasn't reporting errors, (false positives indeed) but enzyme immediately told me what was wrong.

File: gulp/aliases.js [hunk 5]
```diff
@@ -21,8 +21,8 @@ module.exports = (blueprint, gulp) => {
     // perform a full build of the code and then finish
     gulp.task("build", (done) => rs("clean", "compile", "bundle", "webpack-compile-docs", done));
 
-    // build code, run unit tests, terminate
-    gulp.task("test", ["test-dist", "karma", "isotest"]);
+    // run test tasks in series to keep outputs separate
+    gulp.task("test", (done) => rs("test-dist", "karma", "isotest-mocha-w"));
```
Comment (cmslewis) [inline gulp/aliases.js]: Call `done()` here.

**Comment by codebling** [comment]
@giladgray regarding the build failures in [the linked test](https://circleci.com/gh/palantir/blueprint/1508), none of them seem specifically server-rendering related. I can't check out that commit any more because it's been squashed and I see that no changes were made to some of the failing components, so I wonder if it was an issue with the task. I see everything is running well on latest master ([1dced45](https://github.com/palantir/blueprint/tree/1dced4521102bcda0d8d3d33f38c58844569b818)), so you must have figured it out.  Great work !

**Pull request merged**